src/main/java/com/iota/iri/service/ledger/impl/LedgerServiceImpl.java

```diff
@@ -163,54 +162,80 @@ public boolean isBalanceDiffConsistent(Set<Hash> approvedHashes, Map<Hash, Long>
             visitedTransactions.add(solidEntryPointHash);
         });
 
-        final Queue<Hash> nonAnalyzedTransactions = new LinkedList<>(Collections.singleton(startTransaction));
-        Hash transactionPointer;
-        while ((transactionPointer = nonAnalyzedTransactions.poll()) != null) {
-            try {
-                final TransactionViewModel transactionViewModel = TransactionViewModel.fromHash(tangle,
-                        transactionPointer);
-                if (transactionViewModel.getCurrentIndex() == 0 && visitedTransactions.add(transactionPointer)) {
-                    if (transactionViewModel.getType() == TransactionViewModel.PREFILLED_SLOT) {
-                        return null;
-                    }
-                    if (!milestoneService.isTransactionConfirmed(transactionViewModel, milestoneIndex)) {
+        Deque<Hash> stack = new ArrayDeque<>();
+        stack.push(startTransaction);
 
-                        final List<TransactionViewModel> bundleTransactions = bundleValidator.validate(tangle,
-                                snapshotProvider.getInitialSnapshot(), transactionViewModel.getHash());
-
-                        if (bundleTransactions.isEmpty()) {
+        while(stack.size() != 0){
+            try {
+                Map<Hash, Long> currentState = new HashMap<>();
+                Hash transactionPointer = stack.peek();
+                TransactionViewModel transactionViewModel = TransactionViewModel.fromHash(tangle, transactionPointer);
+                TransactionViewModel trunkTransactionViewModel = transactionViewModel.getTrunkTransaction(tangle);
+                TransactionViewModel branchTransactionViewModel = transactionViewModel.getBranchTransaction(tangle);
+
+                boolean approvedTrunk = (trunkTransactionViewModel.snapshotIndex() > 0) && (trunkTransactionViewModel.snapshotIndex() != milestoneIndex);
+                boolean approvedBranch = (branchTransactionViewModel.snapshotIndex() > 0) && (branchTransactionViewModel.snapshotIndex() != milestoneIndex);
+                if ((trunkTransactionViewModel.getType() == TransactionViewModel.PREFILLED_SLOT  || visitedTransactions.contains(trunkTransactionViewModel.getHash()) || approvedTrunk) &&
```

Comment:
Perhaps providing labels for these conditions as booleans instead of rewriting them in the if and else statements here could help to declutter this section. There are a lot of conditions being thrown around very close together here that could make it harder to follow. For example another boolean for each the trunk and branch types to be reused below in the else statement. More descriptive labels or a few extra comments could go a long way towards making the logic easier for others to follow.